Rules/CommonScripts/Spectator.as

```diff
@@ -15,101 +27,188 @@ CPlayer@ targetPlayer()
 	{
 		pos.x += camSpeed;
 		SetTargetPlayer(null);
+		timeToCinematic = CINEMATIC_TIME;
 	}
 	if (controls.ActionKeyPressed(AK_MOVE_UP))
 	{
 		pos.y -= camSpeed;
 		SetTargetPlayer(null);
+		timeToCinematic = CINEMATIC_TIME;
 	}
 	if (controls.ActionKeyPressed(AK_MOVE_DOWN))
 	{
 		pos.y += camSpeed;
 		SetTargetPlayer(null);
+		timeToCinematic = CINEMATIC_TIME;
 	}
 
-    if(controls.isKeyJustReleased(KEY_LBUTTON))
+    if (controls.isKeyJustReleased(KEY_LBUTTON))
     {
         waitForRelease = false;
-
     }
 
-	//Click on players to track them or set camera to mousePos
+	if (timeToCinematic > 0) //player-controlled zoom
+	{
+		if (Maths::Abs(camera.targetDistance - zoomTarget) > 0.001f)
+		{
+			camera.targetDistance = (camera.targetDistance * (3.0f - getRenderApproximateCorrectionFactor() + 1.0f) + (zoomTarget * getRenderApproximateCorrectionFactor())) / 4.0f;
+		}
+		else
+		{
+			camera.targetDistance = zoomTarget;
+		}
+
+		if (AUTO_CINEMATIC)
+		{
+			timeToCinematic -= getRenderApproximateCorrectionFactor();
+		}
+	}
+	else //cinematic camera
+	{
+		//by default, view entire map from center
+		posTarget = Vec2f(mapDim.x, mapDim.y) / 2.0f;
+		f32 zoomW = calculateZoomLevelW(mapDim.x);
+		f32 zoomH = calculateZoomLevelH(mapDim.y);
+		zoomTarget = Maths::Min(zoomW, zoomH);
+		zoomTarget = Maths::Clamp(zoomTarget, 0.5f, 2.0f);
+
+		CBlob@[] blobs;
+		getBlobs(@blobs);
+		calculateImportance(blobs);
+		blobs = sortBlobsByImportance(blobs);
+
+		CBlob@[] players;
+		getBlobsByTag("player", @players);
+
+		if (
+			this.isMatchRunning() && !this.isWarmup() && !this.isGameOver() && //game running
+			!focusOnBlob(blobs) && //not focusing on a blob
+			players.length > 0 //players exist
+		) {
+			u32 maxDistX = 0.0f;
+			u32 maxDistY = 0.0f;
+
+			//calculate mean position of all players
+			posTarget = Vec2f(0, 0);
+			for (uint i = 0; i < players.length; i++)
+			{
+				CBlob@ blob = players[i];
+
+				posTarget += blob.getInterpolatedPosition();
+
+				//look for largest distance between two players
+				for (uint j = 0; j < players.length; j++)
+				{
+					CBlob@ blob2 = players[j];
+					u32 distX = Maths::Abs(blob.getPosition().x - blob2.getPosition().x);
+					u32 distY = Maths::Abs(blob.getPosition().y - blob2.getPosition().y);
+					maxDistX = Maths::Max(distX, maxDistX);
+					maxDistY = Maths::Max(distY, maxDistY);
+				}
+			}
+			posTarget /= players.length;
+
+			if (players.length == 1)
+			{
+				//follow blob with normal zoom
+				zoomTarget = 1.0f;
+			}
+			else
+			{
+				//dynamic zoom to fit all players
+				f32 zoomW = calculateZoomLevelW(maxDistX * 1.8f);
+				f32 zoomH = calculateZoomLevelH(maxDistY * 1.8f);
+				zoomTarget = Maths::Min(zoomW, zoomH);
+			}
+
+			zoomTarget = Maths::Clamp(zoomTarget, 0.75f, 1.5f);
+		}
+
+		//adjust camera pos and zoom
+		camera.targetDistance += (zoomTarget - camera.targetDistance) / CINEMATIC_ZOOM_EASE * getRenderApproximateCorrectionFactor();
```

Comment:
if it turns out that this is really slow on large servers, you could simply calculate the focus/targetPos every few ticks, but update the easing every tick.